// bottender/dist/messenger/MessengerContext.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const invariant_1 = __importDefault(require("invariant"));
const delay_1 = __importDefault(require("delay"));
const warning_1 = __importDefault(require("warning"));
const messaging_api_messenger_1 = require("messaging-api-messenger");
const Context_1 = __importDefault(require("../context/Context"));
class MessengerContext extends Context_1.default {
    constructor({ appId, client, event, session, initialState, requestContext, customAccessToken, batchQueue, emitter, }) {
        super({ client, event, session, initialState, requestContext, emitter });
        this._personaId = null;
        this._customAccessToken = customAccessToken || null;
        this._batchQueue = batchQueue || null;
        this._appId = appId || null;
    }
    get platform() {
        return 'messenger';
    }
    get accessToken() {
        return this._customAccessToken || this._client.accessToken;
    }
    _callClientMethod(method, args) {
        if (this._batchQueue) {
            return this._batchQueue.push(messaging_api_messenger_1.MessengerBatch[method](...args));
        }
        return this._client[method](...args);
    }
    usePersona(personaId) {
        this._personaId = personaId;
    }
    useAccessToken(accessToken) {
        this._customAccessToken = accessToken;
    }
    typing(milliseconds) {
        return __awaiter(this, void 0, void 0, function* () {
            if (milliseconds > 0) {
                yield this.typingOn();
                yield delay_1.default(milliseconds);
                yield this.typingOff();
            }
        });
    }
    sendText(text, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'sendText: should not be called in context without session');
                return;
            }
            if (this._event.isEcho || this._event.isDelivery || this._event.isRead) {
                warning_1.default(false, 'sendText: calling Send APIs in `message_reads`(event.isRead), `message_deliveries`(event.isDelivery) or `message_echoes`(event.isEcho) events may cause endless self-responding, so they are ignored by default.\nYou may like to turn off subscription of those events or handle them without Send APIs.');
                return;
            }
            this._isHandled = true;
            const messagingType = options && 'tag' in options ? 'MESSAGE_TAG' : 'RESPONSE';
            const args = [
                this._session.user.id,
                text,
                Object.assign(Object.assign(Object.assign({ messagingType }, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), (this._personaId ? { personaId: this._personaId } : undefined)), options),
            ];
            return this._callClientMethod('sendText', args);
        });
    }
    getUserProfile(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'getUserProfile: should not be called in context without session');
                return null;
            }
            const args = [
                this._session.user.id,
                Object.assign(Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), options),
            ];
            return this._callClientMethod('getUserProfile', args);
        });
    }
    sendSenderAction(senderAction, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'sendSenderAction: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                senderAction,
                Object.assign(Object.assign(Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), (this._personaId ? { personaId: this._personaId } : undefined)), options),
            ];
            return this._callClientMethod('sendSenderAction', args);
        });
    }
    typingOn(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'typingOn: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                Object.assign(Object.assign(Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), (this._personaId ? { personaId: this._personaId } : undefined)), options),
            ];
            return this._callClientMethod('typingOn', args);
        });
    }
    typingOff(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'typingOff: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                Object.assign(Object.assign(Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), (this._personaId ? { personaId: this._personaId } : undefined)), options),
            ];
            return this._callClientMethod('typingOff', args);
        });
    }
    markSeen(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'markSeen: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                Object.assign(Object.assign(Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), (this._personaId ? { personaId: this._personaId } : undefined)), options),
            ];
            return this._callClientMethod('markSeen', args);
        });
    }
    passThreadControl(targetAppId, metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'passThreadControl: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                targetAppId,
                metadata,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('passThreadControl', args);
        });
    }
    passThreadControlToPageInbox(metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'passThreadControlToPageInbox: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                metadata,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('passThreadControlToPageInbox', args);
        });
    }
    takeThreadControl(metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'takeThreadControl: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                metadata,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('takeThreadControl', args);
        });
    }
    requestThreadControl(metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'requestThreadControl: should not be called in context without session');
                return;
            }
            this._isHandled = true;
            const args = [
                this._session.user.id,
                metadata,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('requestThreadControl', args);
        });
    }
    getThreadOwner() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'getThreadOwner: should not be called in context without session');
                return;
            }
            const args = [
                this._session.user.id,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('getThreadOwner', args);
        });
    }
    isThreadOwner() {
        return __awaiter(this, void 0, void 0, function* () {
            invariant_1.default(this._appId, 'isThreadOwner: must provide appId to use this feature');
            const { appId } = yield this.getThreadOwner();
            return `${appId}` === `${this._appId}`;
        });
    }
    associateLabel(labelId) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'associateLabel: should not be called in context without session');
                return;
            }
            const args = [
                this._session.user.id,
                labelId,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('associateLabel', args);
        });
    }
    dissociateLabel(labelId) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'dissociateLabel: should not be called in context without session');
                return;
            }
            const args = [
                this._session.user.id,
                labelId,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('dissociateLabel', args);
        });
    }
    getAssociatedLabels() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._session) {
                warning_1.default(false, 'getAssociatedLabels: should not be called in context without session');
                return;
            }
            const args = [
                this._session.user.id,
                Object.assign({}, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)),
            ];
            return this._callClientMethod('getAssociatedLabels', args);
        });
    }
}
const sendMethods = [
    ['sendMessage', 3],
    ['sendAttachment', 3],
    ['sendImage', 3],
    ['sendAudio', 3],
    ['sendVideo', 3],
    ['sendFile', 3],
    ['sendTemplate', 3],
    ['sendGenericTemplate', 3],
    ['sendButtonTemplate', 4],
    ['sendMediaTemplate', 3],
    ['sendReceiptTemplate', 3],
    ['sendAirlineBoardingPassTemplate', 3],
    ['sendAirlineCheckinTemplate', 3],
    ['sendAirlineItineraryTemplate', 3],
    ['sendAirlineUpdateTemplate', 3],
];
sendMethods.forEach(([method, arity]) => {
    Object.defineProperty(MessengerContext.prototype, method, {
        enumerable: false,
        configurable: true,
        writable: true,
        value(...args) {
            return __awaiter(this, void 0, void 0, function* () {
                if (!this._session) {
                    warning_1.default(false, `${method}: should not be called in context without session`);
                    return;
                }
                if (this._event.isEcho || this._event.isDelivery || this._event.isRead) {
                    warning_1.default(false, `${method}: calling Send APIs in \`message_reads\`(event.isRead), \`message_deliveries\`(event.isDelivery) or \`message_echoes\`(event.isEcho) events may cause endless self-responding, so they are ignored by default.\nYou may like to turn off subscription of those events or handle them without Send APIs.`);
                    return;
                }
                this._isHandled = true;
                const options = args[arity - 2];
                const messagingType = options && options.tag ? 'MESSAGE_TAG' : 'RESPONSE';
                args[arity - 2] = Object.assign(Object.assign(Object.assign({ messagingType }, (this._customAccessToken
                    ? { accessToken: this._customAccessToken }
                    : undefined)), (this._personaId ? { personaId: this._personaId } : undefined)), options);
                return this._callClientMethod(method, [this._session.user.id, ...args]);
            });
        },
    });
});
exports.default = MessengerContext;
