// messaging-api-line/dist/Line.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const omit_1 = __importDefault(require("lodash/omit"));
function createText(text, options = {}) {
    return Object.assign({ type: 'text', text }, omit_1.default(options, 'accessToken'));
}
function createImage(image, options = {}) {
    return Object.assign({ type: 'image', originalContentUrl: image.originalContentUrl, previewImageUrl: image.previewImageUrl || image.originalContentUrl }, omit_1.default(options, 'accessToken'));
}
function createVideo(video, options = {}) {
    return Object.assign({ type: 'video', originalContentUrl: video.originalContentUrl, previewImageUrl: video.previewImageUrl }, omit_1.default(options, 'accessToken'));
}
function createAudio(audio, options = {}) {
    return Object.assign({ type: 'audio', originalContentUrl: audio.originalContentUrl, duration: audio.duration }, omit_1.default(options, 'accessToken'));
}
function createLocation({ title, address, latitude, longitude }, options = {}) {
    return Object.assign({ type: 'location', title,
        address,
        latitude,
        longitude }, omit_1.default(options, 'accessToken'));
}
function createSticker(sticker, options = {}) {
    return Object.assign({ type: 'sticker', packageId: sticker.packageId, stickerId: sticker.stickerId }, omit_1.default(options, 'accessToken'));
}
function createImagemap(altText, { baseUrl, baseSize, video, actions, }, options = {}) {
    return Object.assign({ type: 'imagemap', baseUrl,
        altText,
        baseSize,
        video,
        actions }, omit_1.default(options, 'accessToken'));
}
function createTemplate(altText, template, options = {}) {
    return Object.assign({ type: 'template', altText,
        template }, omit_1.default(options, 'accessToken'));
}
function createButtonTemplate(altText, { thumbnailImageUrl, imageAspectRatio, imageSize, imageBackgroundColor, title, text, defaultAction, actions, }, options = {}) {
    return createTemplate(altText, {
        type: 'buttons',
        thumbnailImageUrl,
        imageAspectRatio,
        imageSize,
        imageBackgroundColor,
        title,
        text,
        defaultAction,
        actions,
    }, omit_1.default(options, 'accessToken'));
}
function createConfirmTemplate(altText, { text, actions, }, options = {}) {
    return createTemplate(altText, {
        type: 'confirm',
        text,
        actions,
    }, omit_1.default(options, 'accessToken'));
}
function createCarouselTemplate(altText, columns, { imageAspectRatio, imageSize, quickReply, } = {}) {
    return createTemplate(altText, {
        type: 'carousel',
        columns,
        imageAspectRatio,
        imageSize,
    }, { quickReply });
}
function createImageCarouselTemplate(altText, columns, options = {}) {
    return createTemplate(altText, {
        type: 'image_carousel',
        columns,
    }, omit_1.default(options, 'accessToken'));
}
function createFlex(altText, contents, options = {}) {
    return Object.assign({ type: 'flex', altText,
        contents }, omit_1.default(options, 'accessToken'));
}
const Line = {
    createText,
    createImage,
    createVideo,
    createAudio,
    createLocation,
    createSticker,
    createImagemap,
    createTemplate,
    createButtonsTemplate: createButtonTemplate,
    createButtonTemplate,
    createConfirmTemplate,
    createCarouselTemplate,
    createImageCarouselTemplate,
    createFlex,
};
exports.default = Line;
